### PR TITLE
Fix comparing ReferencePositions and update interval collections as positions slide

### PR DESCRIPTION
**Pull request by ransomr**
# Fix interval consistency as positions slide

## Description

This fixes two issues that caused the interval collection index to be inconsistent:
1. LocalReference.getOffset, used to compare ReferencePositions, would always return 0 when on a removed segment. This made it possible for removing a segment to change references that had compared as not equal to change to compare as equal. This broke the red/black tree in interval collections. With this change getOffset may still return a non-zero value when the segment is deleted. This preserves comparison ordering when segments are removed. `Client.localReferencePositionToPosition` has been updated to use an existing method which correctly accounts for removed segments when returning the local position.
2. Added an event emitter on LocalReferences to signal when references slide. Interval collections now listens on these events and updates the index as needed.

Enabled and added tests in `packages\dds\sequence\src\test\intervalCollection.spec.ts`. In addition, improved `assertIntervals` to validate that the index order is as expected (and fixed a bug where it could pass even if intervals were different).

## PR Checklist

-   [X] I have updated the documentation accordingly.
-   [X] I have added tests to cover my changes.
-   [X] All new and existing tests passed.
-   [X] My code follows the code style of this project.
-   [X] I ran the lint checks which produced no new errors nor warnings for my changes.
-   [X] I have checked to ensure there aren't other open Pull Requests for the same update/change.

## Does this introduce a breaking change?

-   [ ] Yes
-   [X] No

> If this introduces a breaking change, please describe the impact and migration path for existing applications below.

## Testing

Tests in `packages\dds\sequence\src\test\intervalCollection.spec.ts`.

### CODE REVIEW

File: packages/dds/sequence/src/intervalCollection.ts [hunk 1]
```diff
@@ -202,7 +202,18 @@ export class Interval implements ISerializableInterval {
     }
 }
 
-export class SequenceInterval implements ISerializableInterval {
+/**
+ * ISequenceIntervalEvents events should only be used internally in IntervalCollections.
```
Comment (Abe27342) [inline packages/dds/sequence/src/intervalCollection.ts:206]:
```suggestion
 * @internal
 * ISequenceIntervalEvents events should only be used internally in IntervalCollections.
```

File: packages/dds/sequence/src/intervalCollection.ts [hunk 2]
```diff
@@ -1209,18 +1298,26 @@ export class IntervalCollection<TInterval extends ISerializableInterval>
...
         if (newStart || newEnd) {
+            // In this case, where we change the start or end of an interval,
+            // it is necessary to remove and re-add the interval listeners.
+            // This ensures that the correct listeners are added to the ReferencePosition.
             this.localCollection.removeExistingInterval(interval);
+            this.localCollection.removeIntervalListeners(interval);
+
             if (newStart) {
                 const props = interval.start.properties;
+                this.client.removeLocalReferencePosition(interval.start);
```
Comment (Abe27342) [inline packages/dds/sequence/src/intervalCollection.ts:1322]: thanks for catching--I had meant to clean this up after noticing it while merging but forgot til just now :)

**Review by Abe27342 — COMMENTED** [review]
Overall looks great. Left a couple small comments.

File: packages/dds/sequence/src/intervalCollection.ts (hunk 1 above)
Comment (Abe27342) [reply packages/dds/sequence/src/intervalCollection.ts:206]: or is this prohibited by api-extractor config? didn't look closely at usage (and IIRC it's somewhat config dependent what it'll let you do in this realm)

Comment (ransomr) [reply packages/dds/sequence/src/intervalCollection.ts:206]: It does cause an api-extractor warning.

Comment (Abe27342) [reply packages/dds/sequence/src/intervalCollection.ts:206]: sounds good, as-is is fine then.

File: packages/dds/sequence/src/intervalCollection.ts [hunk 3]
```diff
@@ -660,7 +731,9 @@ export class LocalIntervalCollection<TInterval extends ISerializableInterval> {
         const newInterval = interval.modify(this.label, start, end, op) as TInterval | undefined;
         if (newInterval) {
             this.removeExistingInterval(interval);
+            this.removeIntervalListeners(interval);
```
[the remark this answers is not in the document]
Comment (ransomr) [reply packages/dds/sequence/src/intervalCollection.ts]: Agree. Fixed.

**Review by Abe27342 — APPROVED** [review]

**Pull request merged**